server/permalinks.go

```diff
@@ -0,0 +1,151 @@
+package main
+
+import (
+	"context"
+	"encoding/base64"
+	"encoding/hex"
+	"fmt"
+	"strings"
+	"time"
+
+	"github.com/xanzy/go-gitlab"
+)
+
+// maxPermalinkReplacements sets the maximum limit to the number of
+// permalink replacements that can be performed on a single message.
+const maxPermalinkReplacements = 10
+
+const permalinkReqTimeout = 5 * time.Second
+
+// maxPreviewLines sets the maximum number of preview lines that will be shown
+// while replacing a permalink.
+const maxPreviewLines = 10
+
+// permalinkLineContext shows the number of lines before and after to show
+// if the link points to a single line.
+const permalinkLineContext = 3
+
+// replacement holds necessary info to replace GitLab permalinks
+// in messages with a code preview block.
+type replacement struct {
+	index         int      // index of the permalink in the string
+	word          string   // the permalink
+	permalinkInfo struct { // holds the necessary metadata of a permalink
```

Comment:
0/5 define a `type permalinkInfo struct` as ^^ and use it in the test, for compactness.

Reply:
Done